### PR TITLE
Changes to how we create blank notebooks

### PR DESCRIPTION
**Pull request by DonJayamanne**
* Changes to how we create blank notebooks
* Added tests
* TODO:
	* File system is not complete (saving, creating new files, save as)
	* Will do separately when using that part of FS

### CODE REVIEW

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -554,13 +555,25 @@ jobs:
         id: test_notebook_vscode
         if: matrix.test-suite == 'notebook'
 
+      - name: Run Remote with VSCode & Jupyter
+        uses: GabrielBB/xvfb-action@v1.4
+        with:
+          run: npm run testRemoteNativeNotebooksInVSCode
```
Comment (DonJayamanne) [inline .github/workflows/pr.yml:561]: Start real jupyter and connect to that.
All `remote` tests will connect & use the same jupyter server. Not going to start & stop for every test suite.

File: src/client/datascience/commands/commandRegistry.ts [hunk 2]
```diff
@@ -50,7 +50,7 @@ export class CommandRegistry implements IDisposable {
         @inject(NotebookCommands) private readonly notebookCommands: NotebookCommands,
         @inject(JupyterCommandLineSelectorCommand)
         private readonly commandLineCommand: JupyterCommandLineSelectorCommand,
-        @inject(INotebookEditorProvider) private notebookEditorProvider: INotebookEditorProvider,
+        @inject(NotebookCreator) private readonly notebookCreator: NotebookCreator,
```
Comment (DonJayamanne) [inline src/client/datascience/commands/commandRegistry.ts:53]: Handling command moved to central location with handling of logic.

File: src/client/datascience/jupyter/kernels/cellExecution.ts [hunk 3]
```diff
@@ -511,7 +511,7 @@ export class CellExecution {
                     password: hasPassword
                 })
                 .then((v) => {
-                    session.sendInputReply(v || '');
+                    session.sendInputReply({ value: v || '', status: 'ok' });
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/cellExecution.ts:514]: Bug fix.

File: src/client/datascience/types.ts [hunk 4]
```diff
@@ -312,7 +312,6 @@ export interface IJupyterPasswordConnect {
 
 export const IJupyterSession = Symbol('IJupyterSession');
 export interface IJupyterSession extends IAsyncDisposable {
-    readonly session: ISessionWithSocket | undefined;
```
Comment (DonJayamanne) [inline src/client/datascience/types.ts:315]: Was not meant to be here.

File: .vscode/launch.json [hunk 5]
```diff
@@ -199,6 +199,7 @@
                 "--extensionTestsPath=${workspaceFolder}/out/test"
             ],
             "env": {
+                "VSC_JUPYTER_ENABLED_EXPERIMENTS": "NativeNotebookEditor,RemoteJupyter", // Comma delimited list of experiments.
```
Comment (DonJayamanne) [inline .vscode/launch.json:202]: Ability to include experiments in tests.

File: src/test/datascience/uiTests/notebookUi.ts [hunk 6]
```diff
@@ -45,7 +46,7 @@ export class NotebookEditorUI extends BaseWebUI {
 
         // Select the cell by focusing it
         const cell = await this.getCell(cellIndex);
-        cell.focus();
+        cell.focus().catch(noop);
```
Comment (DonJayamanne) [inline src/test/datascience/uiTests/notebookUi.ts:49]: Bug

File: src/test/datascience/notebook.functional.test.ts [hunk 7]
```diff
@@ -285,7 +285,7 @@ suite('DataScience notebook tests', () => {
                 // Catch exceptions. Throw a specific assertion if the promise fails
                 try {
                     if (uri) {
-                        ioc.setServerUri(uri);
+                        ioc.setServerUri(uri).catch(noop);
```
Comment (DonJayamanne) [inline src/test/datascience/notebook.functional.test.ts:288]: Bug

File: src/client/remote/ui/notebookCreator.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    private async createBlankNotebookLocally() {
+        await this.notebookEditorProvider.createNew();
+    }
+    private async createBlankOnRemote(connection: JupyterServerConnection) {
+        const remoteFolders = await window.showOpenDialog({
+            canSelectFiles: false,
+            canSelectFolders: true,
+            canSelectMany: false,
+            defaultUri: Uri.file('/').with({ scheme: connection.fileScheme }),
+            openLabel: 'Folder to create notebook',
```
Comment (DonJayamanne) [inline src/client/remote/ui/notebookCreator.ts:42]: TODO:

File: src/client/common/application/commands.ts [hunk 9]
```diff
@@ -92,10 +92,10 @@ export interface ICommandNameArgumentTypeMapping extends ICommandNameWithoutArgu
     ['jupyter.server.logout']: [ServerNode];
     ['jupyter.server.refresh']: [ServerNode];
     ['jupyter.server.kernelSessions.refresh']: [KernelSessionsNode];
-    ['jupyter.server.directory.new']: [FileSystemNode | DirectoryNode];
-    ['jupyter.server.file.new']: [FileSystemNode | DirectoryNode];
-    ['jupyter.server.notebook.new']: [FileSystemNode | DirectoryNode];
-    ['jupyter.server.file.delete']: [FileNode];
+    ['jupyter.server.directory.new']: [FileTreeNodeData | DirectoryTreeNodeData];
```
Comment (DonJayamanne) [inline src/client/common/application/commands.ts:95]: Breaking change in VS Code.

File: src/client/remote/commands.ts [hunk 10]
```diff
@@ -48,34 +49,34 @@ export class CommandRegistry implements IExtensionSingleActivationService {
...
         );
         this.disposables.push(
             this.commandManager.registerCommand('jupyter.server.file.delete', (item) => this.delete(item))
         );
     }
-    private async createNew(item: FileSystemNode | DirectoryNode, type: 'file' | 'directory' | 'notebook') {
-        const remoteFs = await this.fsFactory.getOrCreateRemoteFileSystem(item.data.connection);
-        await remoteFs.createNew(item.uri, type);
-        this.dataProvider.refreshNode(item.data);
+    private async createNew(item: FileTreeNodeData | DirectoryTreeNodeData, type: 'file' | 'directory' | 'notebook') {
```
Comment (DonJayamanne) [inline src/client/remote/commands.ts:60]: VS Code broke it hence changes to the args

File: src/client/common/experiments/service.ts [hunk 11]
```diff
@@ -138,7 +140,13 @@ export class ExperimentService implements IExperimentService {
         if (!this.experimentationService) {
             return;
         }
-
+        if (
+            this.context.extensionMode === ExtensionMode.Test &&
+            process.env.VSC_JUPYTER_ENABLED_EXPERIMENTS &&
+            process.env.VSC_JUPYTER_ENABLED_EXPERIMENTS.includes(experiment)
```
Comment (DonJayamanne) [inline src/client/common/experiments/service.ts:146]: Respect env variable only in tests.

File: src/client/remote/commands.ts [hunk 12]
```diff
@@ -48,34 +49,34 @@ export class CommandRegistry implements IExtensionSingleActivationService {
...
+        this.dataProvider.refreshNode(item);
     }
     private async refreshServer(item: ServerNode) {
         this.dataProvider.refreshNode(item.data);
     }
     private async refreshKernelSessions(item: KernelSessionsNode) {
         this.dataProvider.refreshNode(item.data);
     }
-    private async delete(item: FileNode) {
+    private async delete(item: FileTreeNodeData) {
```
Comment (DonJayamanne) [inline src/client/remote/commands.ts:71]: More changes due to changes in VS Code API (no idea if this change was deliberate, not going to look into that).

**Review by DavidKutu — APPROVED** [review]

File: src/client/remote/ui/fileSystem.ts [hunk 13]
```diff
@@ -183,18 +188,11 @@ export class RemoteFileSystem implements IFileSystemProvider {
...
-                const uri = Uri.file(model.path).with({ scheme: this.scheme });
-                this._fireSoon(
-                    { type: FileChangeType.Created, uri },
-                    { uri: remotePath, type: FileChangeType.Changed }
-                );
-                return uri;
-            }
-        } finally {
-            contentManager.dispose();
+        const model = await contentManager.newUntitled({ type, path: remotePath.fsPath });
```
Comment (rchiodo) [inline src/client/remote/ui/fileSystem.ts:191]: Not sure why you removed the dispose?

**Review by rchiodo — APPROVED** [review]

File: src/client/remote/ui/fileSystem.ts (hunk 13 above)
Comment (DonJayamanne) [reply src/client/remote/ui/fileSystem.ts:191]: Don't want to keep creating that everytime.
Didn't see any point. Basically makes not difference whether i create or destroy.

I figured it was cleaner to re-use same object that trying to be smart & disposing everytime.
When disconnecting from that server, that object gets disposed.

**Pull request merged**